token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs

```diff
@@ -73,54 +76,41 @@ makeTokenLimiter tokenLimiterConfig = do
   tokenLimiterLastServiced <- newMVar (now, min (tokenLimitConfigInitialTokens tokenLimiterConfig) (tokenLimitConfigMaxTokens tokenLimiterConfig))
   pure TokenLimiter {..}
 
--- | Ask if we could debit a number of tokens, without actually doing it.
---
--- Note that this information can become stale _very_ quickly.
--- If you want to also actually debit a number of tokens, use 'tryDebit' instead.
---
--- Warning: this function can block for long if there are other threads waiting to debit tokens.
--- Use in conjunction with an Async.race timeout or similar if you want to avoid lengthy blocking.
-canDebit :: TokenLimiter -> Word64 -> IO Bool
-canDebit TokenLimiter {..} debit = withMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
-  now <- getMonotonicTimeNSec
-  let currentCount = computeCurrentCount tokenLimiterConfig lastServiced countThen now
-  let enoughAvailable = currentCount >= debit
-  pure enoughAvailable
-
 -- | Check if we can debit a number of tokens, and do it if possible.
 --
 -- The returned boolean represents whether the tokens were debited.
 --
--- Warning: blocking caveat described in 'canDebit' applies here too.
+-- Note that there is a small race-condition in which `tryDebit` sometimes
+-- returns `False` eventhough it could (maybe) have debited because another
+-- thread was currently `waitDebit`-ing without actually waiting (because it
+-- didn't need to wait).
 tryDebit :: TokenLimiter -> Word64 -> IO Bool
-tryDebit TokenLimiter {..} debit = modifyMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
-  now <- getMonotonicTimeNSec
-  let currentCount = computeCurrentCount tokenLimiterConfig lastServiced countThen now
-  let enoughAvailable = currentCount >= debit
-  if enoughAvailable
-    then do
-      let newCount = currentCount - debit
-      pure ((now, newCount), True)
-    else pure ((lastServiced, countThen), False)
+tryDebit TokenLimiter {..} debit =
+  fmap (fromMaybe False) $
+    tryModifyMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
+      now <- getMonotonicTimeNSec
+      let currentCount = computeCurrentCount tokenLimiterConfig lastServiced countThen now
+      let enoughAvailable = currentCount >= debit
+      if enoughAvailable
+        then do
+          let newCount = currentCount - debit
+          pure ((now, newCount), True)
+        else pure ((lastServiced, countThen), False)
```

Comment:
optional nit: as both branches end in pure, can pull up the pure in front of the if